File: leetcode/101-symmetric-tree.cpp
/* 101-symmetric-tree.cpp
 *
 * Copyright (C) 2019 Sang-Kil Park <likejazz@gmail.com>
 * All rights reserved.
 *
 * This software may be modified and distributed under the terms
 * of the BSD license.  See the LICENSE file for details.
 */
class Solution {
public:
    bool isSymmetric(TreeNode *root) {
        if (!root)
            return true;

        std::stack<TreeNode *> nodes;
        nodes.push(root->left);
        nodes.push(root->right);

        while (!nodes.empty()) {
            TreeNode *right = nodes.top();
            nodes.pop();
            TreeNode *left = nodes.top();
            nodes.pop();

            if (!left && !right) {
                // Child node does not exist.
                continue;
            } else {
                // Child nodes are not symmetrical.
                if (!left || !right || left->val != right->val) {
                    return false;
                }
            }

            nodes.push(left->right);
            nodes.push(right->left);
            nodes.push(left->left);
            nodes.push(right->right);
        }
        return true;
    }
};

File: python/101-symmetric-tree.py
# 101-symmetric-tree.py
#
# Copyright (C) 2019 Sang-Kil Park <likejazz@gmail.com>
# All rights reserved.
#
# This software may be modified and distributed under the terms
# of the BSD license.  See the LICENSE file for details.
class Solution:
    def __init__(self):
        # Represents tree as array from linked list.
        self.g = [[]]

    def isSymmetric(self, root: TreeNode) -> bool:
        def make_tree_array(node, level):
            # Create empty array if it doesn't exist.
            if len(self.g) < level + 1:
                self.g.append([])
            if node is None:
                self.g[level].append(None)
                return
            else:
                self.g[level].append(node.val)

            make_tree_array(node.left, level + 1)
            make_tree_array(node.right, level + 1)

        make_tree_array(root, 0)

        # Check whether or not is palindrome except root node.
        for g in self.g[1:]:
            while g:
                if g.pop(0) != g.pop(-1):
                    return False
        return True
